### PR TITLE
Delete RuntimeConfigAspect

### PR DESCRIPTION
**Pull request by adamgfraser**
This PR deletes `RuntimeConfigAspect` and the `hook` in `ZIOApp`. Instead, low level functionality like logging is installed using layers and a new set of layer constructors such as `addLogger` are provided to facilitate this. This allows the installation of this functionality to be effectual and resourceful and removes one additional concept.

`RuntimeConfig` currently continues to exist in a limited form to represent a "snapshot" of the current runtime configuration. This is useful in a couple of places.

First, operators such as `succeedWith` currently allow accessing the `RuntimeConfig` and it seems useful here just to represent the "bundle" of the different information that is available.

Second, in mixed applications it facilitates users defining their own `Runtime` without going through a ZIO workflow. This is similar to how `Runtime` contains a `ZEnvironment` even though during execution that will immediately be unpacked into a `FiberRef` value.

### CODE REVIEW

File: core-tests/shared/src/test/scala/zio/ZIOAppSpec.scala [hunk 1]
```diff
@@ -49,7 +49,7 @@ object ZIOAppSpec extends ZIOBaseSpec {
         }
       }
 
-      val app1 = ZIOAppDefault(ZIO.fail("Uh oh!"), RuntimeConfigAspect.addLogger(logger1))
+      val app1 = ZIOApp(ZIO.fail("Uh oh!"), ZLayer.addLogger(logger1))
```
Comment (jdegoes) [inline core-tests/shared/src/test/scala/zio/ZIOAppSpec.scala:52]: Looking at how this reads, I question whether `ZLayer` is the right place for these things.

Yes, they are ZLayer constructors, but they are connected to the concept of runtime system / bootstrapping, not really part of the ZLayer domain (derived and domain-specific versus primitive constructors).

I don't really have a good suggestion but I could imagine creating an `object` like `BootstrapLayers` or `RuntimeLayers` to contain all of them.

File: core/js/src/main/scala/zio/ZIOAppPlatformSpecific.scala [hunk 2]
```diff
@@ -10,13 +10,17 @@ trait ZIOAppPlatformSpecific { self: ZIOApp =>
...
-    val newRuntime     = runtime.mapRuntimeConfig(hook)
-    newRuntime.unsafeRunAsync {
-      invokeWith(newRuntime)(Chunk.fromIterable(args0))
-        .provideEnvironment(runtime.environment)
-        .tapErrorCause(ZIO.logErrorCause(_))
-        .exitCode
-        .tap(exit)
+
+    val newLayer =
+      Scope.default +!+ ZLayer.succeed(ZIOAppArgs(Chunk.fromIterable(args0))) >>>
```
Comment (jdegoes) [inline core/js/src/main/scala/zio/ZIOAppPlatformSpecific.scala:15]: What did we decide about replacing `+++` with `+!+`?

File: docs/datatypes/core/zioapp.md [hunk 3]
```diff
@@ -79,21 +79,10 @@ object CustomizedRuntimeZIOApp extends ZIOAppDefault {
 
 A detailed explanation of the ZIO runtime system can be found on the [runtime](runtime.md) page.
 
-## Installing Low-level Functionalities
```
Comment (jdegoes) [inline docs/datatypes/core/zioapp.md:82]: @khajavi Be sure to incorporate these updates into the docs!

**Review by jdegoes — CHANGES_REQUESTED** [review]
I still think we should delete `RuntimeConfig`. Constructing it will be expensive because it has to be simulated from lots of getter operations across multiple fiber refs; and setting it (not sure if any of these still exist) would not only be expensive, but also poorly defined.

Moreover, if we wanted such functionality, we could extend `FiberRefs` with a phantom type to track _which_ fiber refs it is storing (via some `type Id` inside each `FiberRef`). So you could have stuff like `FiberRefs[currentSupervisor.Id & forkScopeOverride.Id & ...]`. Where a new `get` method would guarantee success in retrieving a value assuming its known the phantom type parameter is tracking that fiber ref. 

From there it's a small step to add a `snapshot` constructor that collects all the fiber refs and bundles them up into a `FiberRefs`. Assuming we want that functionality (and I am not sure if we are using that or if the remaining use cases for `RuntimeConfig` are only ever consuming a part of it).

Open to other views, however, as I haven't had a chance to look at code outside this pull request!

File: core-tests/shared/src/test/scala/zio/ZIOAppSpec.scala (hunk 1 above)
Comment (adamgfraser) [reply core-tests/shared/src/test/scala/zio/ZIOAppSpec.scala:52]: I agree and had the same thought. Another idea would be perhaps they could be moved to the `Runtime` companion object.

File: core/js/src/main/scala/zio/ZIOAppPlatformSpecific.scala (hunk 2 above)
Comment (adamgfraser) [reply core/js/src/main/scala/zio/ZIOAppPlatformSpecific.scala:15]: I think we can replace all of these usages now that we have addressed some of the issues with tags though probably better to do that in a separate PR.

**Comment by adamgfraser** [comment]
I think that is fine and I agree that is cleaner. We just have to think through the implications.

First, that means we don't have a logical grouping for the set of things we return in `succeedWith`. I think we could potentially just change this to give access to the runtime's definition of a fatal error, since this is what is used in low level operators such as `attempt`, and then for anything else you would have to get it separately.

Second, I think that will limit the ability of people to modify the runtime outside of these ZIO operators. For example, creating a custom runtime in a mixed application with a different executor. Perhaps going forward the way you do that is create a ZIO workflow that returns your custom runtime and then unsafely run that workflow with the bootstrap runtime.

File: docs/datatypes/core/zioapp.md (hunk 3 above)
Comment (khajavi) [reply docs/datatypes/core/zioapp.md:82]: Will do!

**Review by jdegoes — APPROVED** [review]

**Pull request merged**